solutions/cf-888E.mdx

```diff
@@ -0,0 +1,136 @@
+---
+id: cf-888E
+source: CF
+title: Maximum Subsequence
+author: Chuyang Wang
+---
+
+[Official Analysis](https://codeforces.com/blog/entry/55701)
+
+## Explanation
+
+For this problem, we are given $n \leq 35$ numbers and have to find a combination such that the sum of all elements in this combination modulo $m$ is maximized. A naive approach, i.e. trying out all $2^{35}$ possible combinations, is not feasible. However, we can use the [Meet-In-The-Middle Technique](https://usaco.guide/gold/meet-in-the-middle) to reduce the time complexity to $2^{\frac{n}{2}}$, which would be enough for our time bound. 
+
+We first split the given array into two halves that are equally long. Then, we generate all combinations of numbers in the first half by using a bit mask and save the results modulo $m$ in a sorted set. For the second half, let's also generate all combinations for them. For each of these combinations, the maximal sum is the sum of all elements in this combination, $current\_sum$, plus the largest sum of a combination of the first half, $left\_sum$, which is less than $m - current\_sum$. It is never optimal to choose a value larger than $m - current\_sum$ since otherwise the result would be $left\_sum + current\_sum - m$, which is strictly less than $current\_sum$. 
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(2^{\lceil \frac{n}{2} \rceil} \cdot n)$
+
+<LanguageSection>
+
+<CPPSection>
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+/**
+ * Calculate the sum of all elements in arr, represented by the binary mask, and
+ * take modulo mod.
+ */
+int unmask(int mask, const vector<int> &arr, int mod) {
+	int current_sum = 0;
+	for (int bit = 0; bit < arr.size(); bit++) {
+		if (mask >> bit & 1 == 1) {
```

Comment:
you sure the order of operators works for this? (just making sure)

Reply:
lol it works (and ac), but not the way I wanted. It does `((mask >> bit) & (1 == 1))` which gives the right answer, but ima gonna add parentheses to make it work "properly"